rslint_core/src/groups/errors/use_isnan.rs

```diff
@@ -0,0 +1,221 @@
+use crate::rule_prelude::*;
+use ast::*;
+use SyntaxKind::*;
+
+declare_lint! {
+    /**
+    Disallow incorrect comparisons against `NaN`.
+
+    `NaN` is a special `Number` value used to represent "not a number" results in calculations.
+    This value is specified in the IEEE Standard for Binary Floating-Point-Arithmetic.
+
+    In JavaScript, `NaN` is unique, it is not equal to anything, including itself! therefore
+    any comparisons to it will either always yield `true` or `false`. Therefore you should
+    use `isNaN(/* num */)` instead to test if a value is `NaN`. This rule is aimed at removing this footgun.
+
+    ## Invalid Code Examples
+
+    ```ignore
+    if (foo == NaN) {
+        // unreachable
+    }
+
+    if (NaN != NaN) {
+        // always runs
+    }
+    ```
+
+    ## Correct Code Examples
+
+    ```ignore
+    if (isNaN(foo)) {
+        /* */
+    }
+
+    if (!isNaN(foo)) {
+        /* */
+    }
+    ```
+    */
+    #[serde(default)]
+    UseIsnan,
+    errors,
+    "use-isnan",
+    /// Switch statements use `===` internally to match an expression, therefore `switch (NaN)` and `case NaN` will never match.
+    /// This rule disables uses like that which are always incorrect (true by default)
+    pub enforce_for_switch_case: bool,
+    /// Index functions like `indexOf` and `lastIndexOf` use `===` internally, therefore matching them against `NaN` will always
+    /// yield `-1`. This option disallows using `indexOf(NaN)` and `lastIndexOf(NaN)` (false by default)
+    pub enforce_for_index_of: bool
+}
+
+impl Default for UseIsnan {
+    fn default() -> Self {
+        Self {
+            enforce_for_switch_case: true,
+            enforce_for_index_of: false,
+        }
+    }
+}
+
+#[typetag::serde]
+impl CstRule for UseIsnan {
+    fn check_node(&self, node: &SyntaxNode, ctx: &mut RuleCtx) -> Option<()> {
+        match node.kind() {
+            BIN_EXPR => {
+                let expr = node.to::<BinExpr>();
+                if !expr.comparison() {
+                    return None;
+                }
+
+                let opposite = if expr.lhs().filter(|e| e.text() == "NaN").is_some() {
+                    expr.rhs()?
+                } else if expr.rhs().filter(|e| e.text() == "NaN").is_some() {
+                    expr.lhs()?
+                } else {
+                    return None;
+                };
+                let op = expr.op().unwrap();
+
+                let always_text = if matches!(op, op!(!=) | op!(!==)) {
+                    "true"
+                } else {
+                    "false"
+                };
+
+                let mut err = ctx
+                    .err(
+                        self.name(),
+                        format!(
+                            "comparing `{}` to `NaN` using `{}` will always return {}",
+                            opposite.text(),
```

Comment:
I would do the same here like in valid-typeof, so we only print the value if it's shorter than 20 chars. This would also open the possibility to create a helper method or Macro to only return the full string if it's shorter than 20 chars.